pup-0006.md

```diff
@@ -0,0 +1,101 @@
+---
+title: Commit Bit Assignment
+author: Brian Bouterse
+created: 01-06-2018
+status: Approved
+---
+
+## Summary
+
+For pulpcore and its related tools, there is no written process to describe giving the commit bit to
+a contributor for all or a portion of the codebase. The existing committers want to openly document
+the process.
+
+
+## Motivation
+
+Historically the commit bit was given on day 0 when hired on the core team at Red Hat.  In Oct 2017
+the existing committers decided to stop that practice and instead document an open process.
+Engineers hired on the Pulp team since Oct '17 have not received commit bit.
+
+We have not yet documented an open process of which to give a proven contributor commit access to
+all or a portion of the codebase.
+
+
+## Scope
+
+This formally applies to the following core repositories and core-team maintained plugins:
+
+Core Repos:
+* pulp/pulp
+* pulp/devel
+* pulp/pulp-ci
+
+Plugin Repos:
+* pulp/pulp_file
+* pulp/pulp_python
+* pulp/pulp_rpm
+* pulp/crane
+* pulp/pulp_docker
+* pulp/pulp_puppet
+* pulp/pulp_ostree
+
+This policy is intended as the default for all software in the Pulp ecosystem unless explicitly
+stated otherwise.
+
+
+## Criteria
+
+Overall the candidate must have demonstrated commitment and care towards the needs of all Pulp
+users and not only their own interests. Also they must have the experience to be trusted with major
+aspects of Pulp functionality in that area.
+
+These requirements are somewhat vague by design and leave the decision to the judgement of the
+existing developers. Users rely on the judgement of developers already, so to rely on that judgement
+to accept new contributors is reasonable. Anyone who specifically wants to get more involved should
+approach the core devs about mentorship.
+
+
+## Process
+
+A nomination or self-nomination can be sent to the pulp-dev email list and must contain state the
+following:
+
+* Name and email of the nominee. This can be a self-nomination.
+* Repositories being requested
+* [Specific paths](https://blog.github.com/2017-07-06-introducing-code-owners/) or the entire repo
```

Comment:
-1 on specific paths, i have a feeling it would introduce a lot of confusion.
say there is a PR for path A, B, C.
core reviewer has commit bit for A,C,D. he would review this PR but won't be able to merge.

Candidate applies for multiple paths commit bit acquisition, how the voting will go in case there will be different commit holders for different paths?

Reply:
say we've got docs people, CI people and code people, and say I've just sent a PR touching docs, CI and the code, how the PR going to be merged? Shall I go to every group and pull them for reviews?

How about wikis, issues etc?

Reply:
I suspect that some people will have ownership of more than one are of code. Some people may even have ownership of the entire repository. I would expect the docs to be owned by everyone who has a commit bit. 

I doubt that the situation you are describing will be very common. In situations where a review is needed from 2 people in order to merge, we will have even more confidence that the code is introducing the intended changes.

Reply:
I'm sorry but it sounds like an unnecessary complication to me since I lack the knowledge of the motivation behind such a decision. If it indeed is the case that we objectively need the ability to selectively grant commit permissions to particular repository regions, we'd better clearly state reasons; although I can't prove it anyhow, this splitting doesn't seem a common practise in the communities.

Reply:
But the commit bit isn't as much about the knowledge as about the trust: we trust the folks to do their best to approve or disapprove PRs.
Moreover asking other folks for a PR helps to spread the knowledge.
Also mind @asmacdo might shift focus one day, should the commit bit for a subdirectory follow his focus?
I'm sorry this doesn't seem a sufficient reason to me to accept the directory commit permissions.

Reply:
I like the concept of the idea. I'd love to be able to review all the PRs that use DRF. But I do have some concerns.

# I agree with Milan's concern. I think the need for 2 reviewers seems like it would happen often, if only for an import statement change.
# How could we make it obvious who needed to review each chunk?
# PR review is one of the ways to learn an area we don't know as well, particularly when reviewing code written by one of the area specialists.
# For reviewers with more limited permissions, PR authors have a disincentive to ask for review, but we want to encourage building up the review skills of newer committers.

If most of the review is done by the people who do review can review everything, we will avoid most of these concerns. The people with commit bit for the entire repo should be limited, but should still include everyone who has been working on it full time for a while. New or infrequent committers might be given access to particular areas.